### PR TITLE
Rascal to iguana grammar conversion

### PR DESCRIPTION
**Pull request by afroozeh**
This PR is the first step to provide support for Iguana in Rascal. This PR converts the Rascal grammar into an Iguana grammar.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -50,10 +50,15 @@
...
-                    <version>0.24.4</version>
-                </dependency>
-	</dependencies>
+        <dependency>
+            <groupId>org.rascalmpl</groupId>
+            <artifactId>rascal</artifactId>
+            <version>0.24.4</version>
+        </dependency>
+        <dependency>
+            <groupId>iguana</groupId>
```
Comment (afroozeh) [inline pom.xml:59]: To get the Iguana dependency, clone the Iguana project and run `mvn clean install`. That would put the jar into the local maven repo.

File: src/util/Iguana.rsc [hunk 2]
```diff
@@ -3,15 +3,16 @@ module util::Iguana
...
 
-syntax A = "a";
+// import lang::rascal::\syntax::Rascal;
+import demo::lang::Pico::Syntax;
 
 alias Parser[&T] = &T (str input, loc origin);
 
 @javaClass{util.ParserGenerator}
-java Parser[&T] parser(type[&T] grammar); 
+java Parser[&T] createParser(type[&T] grammar);
```
Comment (afroozeh) [inline src/util/Iguana.rsc:12]: I had to rename this method, somehow there was a conflict and Rascal was trying to load the default parse method.

**Comment by jurgenvinju** [comment]
This looks like a good baseline to work from!

**Comment by afroozeh** [comment]
@jurgenvinju thanks! I'll start with ParseTree conversion and open issues that we need to discuss, for example, how to handle parameterized sorts in Iguana.

**Comment by jurgenvinju** [comment]
Ok. In principle, parameterized sorts in Rascal are a pre-processor feature. So when you get a parametrized sort in Iguana it is always already bound and the complex symbol `MySort[MyParameter]` could just as well be read as `MySort_MyParameter_`. I.e. it is just a unique name for a non-terminal. 

To get from a grammar that still has open non-terminals there is a library function that binds them and produces a grammar with only closed non-terminals. It's somewhere in `lang::rascal::grammar::definition`

**Comment by afroozeh** [comment]
That's perfect, so I can assume that I create a nonterminal in Iguana for the sort `MySort[MyParameter]`. I'll push a commit to have this and then we can have the full Rascal grammar.

**Comment by afroozeh** [comment]
@jurgenvinju I'm a bit confused about the conversion between types and grammars in Rascal.
The current `createParser` method is defined like this;
```
java Parser[&T] createParser(type[&T] grammar);
```
Assuming the start syntax of a grammar is S, then I can get a grammar like this (I hope it's correct, if not please correct it):
```
 Grammar g = expandParameterizedSymbols(grammar({#S.symbol}, #S.definitions));
```
But when I call `createParser(g)` it throws an error because of incompatible types which is expected. How do I get the type back from `g`? Or should we change the method signature to operate with grammars?

**Comment by jurgenvinju** [comment]
* You can create a grammar like this: `Grammar g = grammar(#S)`, but it will simply do what you did. 
* If you need to recreate the reified type from a grammar you can use `type(symbol, definitions)`, so `#S == type(#S.symbol, #S.definitions)`.
* In Java you can use the `IRascalValueFactory.reifiedType` factory methods for creating a reified type from a symbol and a map of non-terminal definitions

**Comment by afroozeh** [comment]
I think I'm not getting this part. Basically I want to expand the parametrized types before calling iguana. This is the signature of the `createParser` method:
```
java Parser[&T] createParser(type[&T] grammar);
```
Let's say we have this grammar:
```
syntax A = "a";
```
Then how should I go from the expanded grammar to `type[&T]` when calling the method? In the example above, you mentioned `#S == type(#S.symbol, #S.definitions)` but I want to get the symbol for the expanded grammar. What am I missing here?
```
Grammar g = grammar(#A);
Grammar expandedGrammar = expandParameterizedSymbols(g);
createParser(???);
```

**Comment by jurgenvinju** [comment]
Thanks for fixing this yourself!

**Pull request merged**